File: javascript/bob/bob.js
var Bob = function () {}

Bob.prototype.hey = function (input) {
  if (isEmpty(input)) return "Fine. Be that way!"
  if (isYelling(input)) return "Whoa, chill out!"
  if (isQuestion(input)) return "Sure."
  return "Whatever."
}

const isEmpty = input => input.trim() === ""
const isQuestion = input => input.endsWith('?')
const isYelling = input => (input.toUpperCase() === input) && (input.toLowerCase() !== input)

module.exports = Bob

File: haskell/bob/src/Bob.hs
module Bob (responseFor) where

import Data.Char (toUpper, toLower, isSpace)

data Response
    = Question
    | Yelling
    | Silence
    | Whatever

responseFor :: String -> String
responseFor = respondTo . determinePhrase 

respondTo :: Response -> String
respondTo r =
    case r of
        Question -> "Sure."
        Yelling -> "Whoa, chill out!"
        Silence -> "Fine. Be that way!"
        Whatever -> "Whatever."

determinePhrase :: [Char] -> Response
determinePhrase p =
    let trimmed = filter (not . isSpace) p
    in
    if (trimmed == []) then Silence 
    else if ((fmap toUpper trimmed == trimmed) && (fmap toLower trimmed /= trimmed)) then Yelling
    else if ((head $ reverse trimmed) == '?') then Question
    else Whatever

